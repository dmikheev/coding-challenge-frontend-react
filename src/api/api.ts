interface IApiResponseResult {
  incidents: IIncident[];
  total: number;
}

interface IRawResponse {
  incidents: IIncident[];
}
interface IIncident {
  id: number;
  title: string;
  description: string;
  address: string;
  occurred_at: number;
  updated_at: number;
  source: {
    html_url: string;
  };
  media: {
    image_url_thumb: string;
  };
}

const PROXY_URL = 'https://cors-anywhere.herokuapp.com/';
const ORIGINAL_URL = 'https://bikewise.org:443/api/v2/incidents';

const PER_PAGE = 10;
const INCIDENT_TYPE = 'theft';
const PROXIMITY = 'Berlin, DE';
const PROXIMITY_SQUARE = 50;

function getUrl(page: number): string {
  const params = {
    page,
    per_page: PER_PAGE,
    incident_type: INCIDENT_TYPE,
    proximity: encodeURIComponent(PROXIMITY),
    proximity_square: PROXIMITY_SQUARE,
  };
  const paramsStr = (Object.keys(params) as Array<keyof typeof params>)
    .map((paramName) => `${paramName}=${params[paramName]}`)
    .join('&');

  return `${PROXY_URL}${ORIGINAL_URL}?${paramsStr}`;
}

const api = {
  getIncidents(page = 1): Promise<IApiResponseResult> {
    return fetch(getUrl(page))
      .then((res) => {
        if (!res.ok) {
          throw res;
        }

        return res.json()
          .then((responseJson: IRawResponse) => {
            return {
              incidents: responseJson.incidents,
              total: Number(res.headers.get('total')),
            };
          });
      });
  },
};
export default api;
