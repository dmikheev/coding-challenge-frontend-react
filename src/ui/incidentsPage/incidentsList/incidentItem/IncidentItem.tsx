import { Link, Paper, Typography } from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import moment from 'moment';
import React from 'react';
import styled, { css } from 'styled-components';
import { IIncident } from '../../../../api/api';
import bikePhotoPlaceholder from './bike_photo_placeholder.svg';

interface IIncidentItemProps {
  className?: string;
  data: IIncident;
}
const IncidentItem: React.FC<IIncidentItemProps> = ({ className, data }) => {
  const descriptionHtml = !!data.description && (
    <StyledDescriptionWrap>
      {data.description}
    </StyledDescriptionWrap>
  );

  const occurredDate = moment(data.occurred_at).format('LL');
  const reportedDate = moment(data.updated_at).format('LL');
  const reportedDateHtml = (occurredDate !== reportedDate) && ` (Reported on ${reportedDate})`;

  return (
    <StyledWrap className={className}>
      <StyledPhoto
        isPlaceholder={!data.media.image_url_thumb}
        src={data.media.image_url_thumb || bikePhotoPlaceholder}
        alt="bike"
      />
      <div>
        <StyledTitleWrap>
          <Link variant="h5" href={data.source.html_url} target="_blank" rel="noopener">
            {data.title}
          </Link>
        </StyledTitleWrap>
        {descriptionHtml}
        <Typography>{data.address} - {occurredDate} {reportedDateHtml}</Typography>
      </div>
    </StyledWrap>
  );
};
export default IncidentItem;

const StyledWrap = styled(Paper)`
  display: flex;
  align-items: flex-start;
  padding: 20px;
`;

interface IStyledPhotoProps {
  isPlaceholder: boolean;
}
const photoSize = '250px';
const StyledPhoto = styled.img<IStyledPhotoProps>`
  flex-shrink: 0;
  margin-right: 20px;
  width: ${photoSize};
  height: ${photoSize};
  background-color: ${grey[300]};
  
  ${props => props.isPlaceholder && css`
    padding: 20px;
  `}
`;

const StyledTitleWrap = styled.div`
  margin-bottom: 10px;
`;

const StyledDescriptionWrap = styled(Typography)`
  margin-bottom: 10px;
`;
