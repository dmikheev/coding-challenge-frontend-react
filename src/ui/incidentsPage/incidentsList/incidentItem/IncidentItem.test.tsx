import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { IIncident } from '../../../../api/api';
import IncidentItem from './IncidentItem';

const testIncidentData: IIncident = {
  id: 123,
  title: 'Test title',
  description: 'Detailed test description',
  address: 'Test address',
  occurred_at: (new Date(2019, 0, 1)).getTime(),
  updated_at: (new Date(2019, 1, 2)).getTime(),
  source: {
    html_url: 'https://test.com/bike',
  },
  media: {
    image_url_thumb: 'https://test.com/bike/photo.jpg',
  },
};

it('renders correctly and shows incident data', () => {
  const wrapper = shallow(
    <IncidentItem className="test" data={testIncidentData}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders correctly and shows placeholder image if photo url is empty', () => {
  const incidentData: IIncident = {
    ...testIncidentData,
    media: {
      image_url_thumb: '',
    },
  };
  const wrapper = shallow(
    <IncidentItem className="test" data={incidentData}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
