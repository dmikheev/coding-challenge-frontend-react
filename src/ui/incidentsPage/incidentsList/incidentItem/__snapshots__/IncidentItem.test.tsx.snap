// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly and shows incident data 1`] = `
<Styled(WithStyles(ForwardRef(Paper)))
  className="test"
>
  <styled.img
    alt="bike"
    isPlaceholder={false}
    src="https://test.com/bike/photo.jpg"
  />
  <div>
    <styled.div>
      <WithStyles(ForwardRef(Link))
        href="https://test.com/bike"
        rel="noopener"
        target="_blank"
        variant="h5"
      >
        Test title
      </WithStyles(ForwardRef(Link))>
    </styled.div>
    <Styled(WithStyles(ForwardRef(Typography)))>
      Detailed test description
    </Styled(WithStyles(ForwardRef(Typography)))>
    <WithStyles(ForwardRef(Typography))>
      Test address
       - 
      January 1, 2019
       
       (Reported on February 2, 2019)
    </WithStyles(ForwardRef(Typography))>
  </div>
</Styled(WithStyles(ForwardRef(Paper)))>
`;

exports[`renders correctly and shows placeholder image if photo url is empty 1`] = `
<Styled(WithStyles(ForwardRef(Paper)))
  className="test"
>
  <styled.img
    alt="bike"
    isPlaceholder={true}
    src="bike_photo_placeholder.svg"
  />
  <div>
    <styled.div>
      <WithStyles(ForwardRef(Link))
        href="https://test.com/bike"
        rel="noopener"
        target="_blank"
        variant="h5"
      >
        Test title
      </WithStyles(ForwardRef(Link))>
    </styled.div>
    <Styled(WithStyles(ForwardRef(Typography)))>
      Detailed test description
    </Styled(WithStyles(ForwardRef(Typography)))>
    <WithStyles(ForwardRef(Typography))>
      Test address
       - 
      January 1, 2019
       
       (Reported on February 2, 2019)
    </WithStyles(ForwardRef(Typography))>
  </div>
</Styled(WithStyles(ForwardRef(Paper)))>
`;
